Created a mailer to handle the Adoption Application forms.

JessSumner opened the pull request:
Why:
When a form is completed and submitted from the website the rescue needs to be notified and have access to the data.

This PR:
Generates a mailer for the Adoption Application and includes a method to send an email.
Has added securely an admin email that can be called without making this information public.
Corrected issues with the radio buttons not actually saving an input to the database.
Has created an HTML and text template for the email.
Includes a preview of the HTML email.

JessSumner committed:
Generated a mailer for adoption application forms

Why:
When an adoption application form is submitted I want the form to be emailed to the rescue.

JessSumner committed:
Added html view and partial for the email content

The data from the application form has to be presented in a logical manner for the rescue to process.

The data is stored in a table.

JessSumner committed:
Preview HTML Mailer working for adoption application

The new mailer renders a partial for the form content which in turn renders a partial for each data entry.
This way the formating for table is defined only once in the data entry partial.
Originally the method for sending the email took in an argument called object. This has been renamed to be adoption_application. This is for clarity purposes.

JessSumner committed:
Styled the HTML Email content

Added some simple inline styles to the html table.
Added border around each section. Width is auto adjusting for the answer section so the table will be uniform based on the longest answer.

JessSumner committed:
Formatted html table and renamed the partial to be more descriptive.

The second partial rendered is now called table row as it renders the question and answer for each row individually.
Table formatted so there is no space between each cell.

JessSumner committed:
Updated to add an admin email.

Used a config variable to save an Admin email.
The completed forms should mail out to this email.

JessSumner committed:
Updated Radio Buttons for Adoption Application form

Why:
I noticed that when the data saved to the model there was no value saved for the questions asked using radio buttons. The emails being sent to the rescue therefore just had a blank cell for these questions.

This commit:
Changes the radio_button_tag to be f.radio_button. This links the radio button to the form and as a result the value chosen is saved with all the other inputs from the other form fields.

Side effects:
The way the arguments and options are written needed to be changed based on this.
Yes and no replaced with true and false.
Instead of writing just true, now have to include :checked => true in order to default select one option.
This caused the dependent fields to break as the data-radio-name now needs to include the form name.

JessSumner committed:
Added a text mailer format.

Maintained the same file structure as the HTML mailer but removed the html styling.

JessSumner committed:
Email called to send

Up until now I had not actually called the email method to send.
Now if the adoption application is valid and saved and email will send on submission.
Corrected the preview to call the first adoption application rather than the last.

JessSumner commented:
@MattMSumner @halogenandtoast @kenyonj I'm less confident in this PR as I covered several new topics. If you have a few moments I would greatly appreciate your feedback. Thank you

halogenandtoast commented on app/mailers/adoption_application_mailer.rb:
```diff
@@ -0,0 +1,12 @@
+class AdoptionApplicationMailer < ApplicationMailer
+  def new_adoption_application(adoption_application)
```
I think I would rename both the class and this method. It seems like this is more contextual than your class implies. For instance I would choose `NewAdoptionApplicationMailer` which describes better what it's purpose is (to send mails when a new application is created). Then I'd probably just call my method `notification`. What do you think?

halogenandtoast commented on app/views/adoption_application_mailer/_table_row.html.erb:
```diff
@@ -0,0 +1,4 @@
+<tr>
+  <td style="border:solid; border-width:1px; font-weight:bold;"><%= data_entry.to_s %></td>
```
`to_s` isn't necessary here as <%= thing %> calls `to_s` on `thing` automatically.

halogenandtoast commented on app/views/adoption_application_mailer/_text_table_row.text.erb:
```diff
@@ -0,0 +1 @@
+<%= data_entry.to_s %>: <%= adoption_application.public_send(data_entry) %>
```
`to_s` isn't needed here either.

halogenandtoast commented on app/views/adoption_application_mailer/new_adoption_application.html.erb:
```diff
@@ -0,0 +1,14 @@
+  <body>
+    <h1 style="color:#363691">A new adoption application form has been submitted.</h1>
+    <div>
+      <%= render "adoption_application_mailer_content",
```
I don't think a partial was necessary here.

halogenandtoast commented on app/views/adoption_application_mailer/new_adoption_application.text.erb:
```diff
@@ -0,0 +1,4 @@
+A new adoption application form has been submitted.
+
+<%= render "text_adoption_application_mailer_content",
```
I don't think a partial was necessary here.

halogenandtoast commented on app/views/adoption_applications/_form_fields.html.erb:
```diff
@@ -23,14 +23,17 @@
-    <%= radio_button_tag(:current_rats, "yes") %>
-    <%= label_tag(:current_rats_yes, "I currently own one or more rats") %></br>
+    <%= f.radio_button :current_rats, true %>
+    <%= f.label :current_rats, "I currently own one or more rats", :value => "true" %></br>
```
I'd suggest using the newer hash syntax

``` erb
<%= f.label :current_rats, "I currently own one or more rats", value: "true" %>
```

halogenandtoast replied:
I've never used the `value` option before. Pretty neat!

halogenandtoast commented on app/views/adoption_applications/_form_fields.html.erb:
```diff
@@ -23,14 +23,17 @@
 
-    <%= radio_button_tag(:current_rats, "no", true) %>
-    <%= label_tag(:current_rats_no, "I do not currently own any rats") %>
+    <%= f.radio_button :current_rats, false, :checked => true %>
```
Same here with the hash syntax.

halogenandtoast replied:
I've never used the `checked` option before. Pretty neat!

halogenandtoast commented on app/views/adoption_applications/_form_fields.html.erb:
```diff
@@ -23,14 +23,17 @@
-    <%= label_tag(:current_rats_no, "I do not currently own any rats") %>
+    <%= f.radio_button :current_rats, false, :checked => true %>
+    <%= f.label :current_rats, "I do not currently own any rats",
+                :value => "false" %>
```
Same here with the hash syntax.

halogenandtoast commented on app/views/adoption_application_mailer/_adoption_application_mailer_content.html.erb:
```diff
@@ -0,0 +1,50 @@
+<table style="border-collapse: collapse;">
+  <%= render "table_row", adoption_application: adoption_application,
```
Not sure how I feel about this partial usage. It's still a lot of repetition. You could do something like.

``` ruby
class AdoptionApplication
  PUBLIC_FIELDS =  %i(name address city cell state zip email current_rates
      current_rat_count current_rat_gender current_rat_age previous rates
      previous_rate_count previous_rat_lifespan cause_of_death vet_care
      vet_name  cage_description cage_location household_members
      children_ages additional_animals interactions_time preferred_characteristics)
```

then here maybe

``` erb
<% AdoptionApplication::PUBLIC_FIELDS.each do |field| %>
  <tr>
    <td style="border:solid; border-width:1px; font-weight:bold;"><%= field %></td>
    <td style="border:solid; border-width:1px; width: auto;"><%= @adoption_application.attributes[field] %></td>
  </tr>
<% end %>
```

halogenandtoast replied:
That avoids the public send, and can be reused in the text version as well.

JessSumner replied:
@halogenandtoast This suggestion is exactly what I hoped was possible but couldn't see how to do. I've changed it slightly but thank you so much for the guidance.

JessSumner committed:
Corrections and refactors

Improved the mailer views. The form fields have been pulled into a constant in the model. This constant is an array of strings. This can then be called by the mailer view and looped over. Each string can be used for the left column and then used to call the corresponding value from the adoption application. The html formatting can be called within the loop. This cuts out the use of two partials.

Updated my hash syntax to not use the =>
Removed some instances of .to_s as <%= %> does that anyway.
Renamed the mailer NewAdoptionApplicationMailer as it is handling a new adoption application and the email method is notification.
Shortened a few lines below the 80 character limit and replaced single quotes with double quotes.

JessSumner committed:
Freeze constant and remove whitespace

Frozen the constant array. Indented the subsequent lines of the array.
Removed whitespace from the mailer method.

JessSumner committed:
Moved each item of the constant array to it's own line

Style guide suggestion.
